lib/daru/vector.rb

```diff
@@ -1481,13 +1481,16 @@ def prepare_bootstrap(estimators)
 
     # Note: To maintain sanity, this _MUST_ be the _ONLY_ place in daru where the
     # @dtype variable is set and the underlying data type of vector changed.
-    def cast_vector_to dtype, source=nil, nm_dtype=nil
+    def cast_vector_to dtype, source=nil, nm_dtype=nil # rubocop:disable Style/CyclomaticComplexity
       source = @data.to_a if source.nil?
 
       new_vector =
         case dtype
         when :array   then Daru::Accessors::ArrayWrapper.new(source, self)
-        when :nmatrix then Daru::Accessors::NMatrixWrapper.new(source, self, nm_dtype)
+        when :nmatrix then
+          # To avoid arrays with nils throwing TypeError for :int32 default nm_dtype
+          nm_dtype = :object if nm_dtype.nil? && source.any?(&:nil?)
```

Comment:
@zverok yes that's what I mean. If you don't pass `nm_dtype` at all then it will use `:int32`  so we would check for nil values in the vector. But if you explicitly pass `:int32` as your `nm_dtype` and you have nils in the vector, you'd expect it to fail right? Because checking for nils in that case would be unnecessarily inefficient. That's the scenario I'm trying to avoid.